tests/ci/integration/python_patch/main/aws-lc-cpython.patch

```diff
@@ -0,0 +1,555 @@
++    -l:libcrypto.a -Wl,--exclude-libs,libcrypto.a
+ 
+ # The _tkinter module.
+ #
+diff --git a/Modules/_hashopenssl.c b/Modules/_hashopenssl.c
+index 0e230f332f..390d6e0ed9 100644
+--- a/Modules/_hashopenssl.c
++++ b/Modules/_hashopenssl.c
+@@ -130,8 +130,12 @@ static const py_hashentry_t py_hashes[] = {
+     PY_HASH_ENTRY(Py_hash_shake_128, NULL, SN_shake128, NID_shake128),
+     PY_HASH_ENTRY(Py_hash_shake_256, NULL, SN_shake256, NID_shake256),
+     /* blake2 digest */
++#if defined(NID_blake2s256)
+     PY_HASH_ENTRY(Py_hash_blake2s, "blake2s256", SN_blake2s256, NID_blake2s256),
++#endif
++#if defined(NID_blake2b512)
+     PY_HASH_ENTRY(Py_hash_blake2b, "blake2b512", SN_blake2b512, NID_blake2b512),
++#endif
+     PY_HASH_ENTRY(NULL, NULL, NULL, 0),
+ };
+ 
+diff --git a/Modules/_ssl.c b/Modules/_ssl.c
+index bc30290942..e0e59b97e5 100644
+--- a/Modules/_ssl.c
++++ b/Modules/_ssl.c
+@@ -187,6 +187,13 @@ extern const SSL_METHOD *TLSv1_2_method(void);
+ #endif
+ 
+ 
++
++#if defined(OPENSSL_IS_AWSLC) || !defined(TLS1_3_VERSION) || defined(OPENSSL_NO_TLS1_3)
++  #define PY_SSL_NO_POST_HS_AUTH
++  #define OPENSSL_NO_PSK
++#endif
++
++
+ enum py_ssl_error {
+     /* these mirror ssl.h */
+     PY_SSL_ERROR_NONE,
+@@ -231,7 +238,7 @@ enum py_proto_version {
+     PY_PROTO_TLSv1 = TLS1_VERSION,
+     PY_PROTO_TLSv1_1 = TLS1_1_VERSION,
+     PY_PROTO_TLSv1_2 = TLS1_2_VERSION,
+-#ifdef TLS1_3_VERSION
++#if defined(TLS1_3_VERSION)
+     PY_PROTO_TLSv1_3 = TLS1_3_VERSION,
+ #else
+     PY_PROTO_TLSv1_3 = 0x304,
+@@ -293,7 +300,7 @@ typedef struct {
+      */
+     unsigned int hostflags;
+     int protocol;
+-#ifdef TLS1_3_VERSION
++#if !defined(PY_SSL_NO_POST_HS_AUTH)
+     int post_handshake_auth;
+ #endif
+     PyObject *msg_cb;
+@@ -885,7 +892,7 @@ newPySSLSocket(PySSLContext *sslctx, PySocketSockObject *sock,
+     SSL_set_mode(self->ssl,
+                  SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER | SSL_MODE_AUTO_RETRY);
+ 
+-#ifdef TLS1_3_VERSION
++#if !defined(PY_SSL_NO_POST_HS_AUTH)
+     if (sslctx->post_handshake_auth == 1) {
+         if (socket_type == PY_SSL_SERVER) {
+             /* bpo-37428: OpenSSL does not ignore SSL_VERIFY_POST_HANDSHAKE.
+@@ -1028,6 +1035,7 @@ _ssl__SSLSocket_do_handshake_impl(PySSLSocket *self)
+     } while (err.ssl == SSL_ERROR_WANT_READ ||
+              err.ssl == SSL_ERROR_WANT_WRITE);
+     Py_XDECREF(sock);
++
+     if (ret < 1)
+         return PySSL_SetError(self, ret, __FILE__, __LINE__);
+     if (PySSL_ChainExceptions(self) < 0)
+@@ -2060,7 +2068,12 @@ _ssl__SSLSocket_shared_ciphers_impl(PySSLSocket *self)
+     len = 0;
+     for (i = 0; i < sk_SSL_CIPHER_num(server_ciphers); i++) {
+         cipher = sk_SSL_CIPHER_value(server_ciphers, i);
++#if defined(OPENSSL_IS_AWSLC)
++        size_t unused_idx;
++        if (sk_SSL_CIPHER_find(client_ciphers, &unused_idx, cipher) < 0)
++#else
+         if (sk_SSL_CIPHER_find(client_ciphers, cipher) < 0)
```

Comment:
Why not add a find that doesn't take an out index that wraps the existing sk_find for better compatibility? Also can our version of find ever return a negative number?

Reply:
> Why not add a find that doesn't take an out index that wraps the existing sk_find for better compatibility?

I don't think we can do this, as (to my knowledge) C doesn't support function overloading. Or are you proposing that we create a new funciton/symbol altogether that wraps our existing `sk_SSL_CIPHER_find` implementation and use that here?

I suppose we could also just [change the signature to match OpenSSL's](https://github.com/aws/aws-lc/blob/main/include/openssl/stack.h#L202-L204), but that might risk breaking any consumers that already use that function.

> Also can our version of find ever return a negative number?

Not [according to our documentation](https://github.com/aws/aws-lc/blob/main/include/openssl/stack.h#L199-L200), no. Good catch! I'll update this.

Reply:
I was thinking we could have `sk_SSL_CIPHER_find_internal` which is our current implementation and `sk_SSL_CIPHER_find` which matches OpenSSL. I don't have a sense for how common OpenSSL stack usage outside of libcrypto/libssl is.

Reply:
> I don't have a sense for how common OpenSSL stack usage outside of libcrypto/libssl is.

It looks like it's [non-zero](https://github.com/search?q=sk_SSL_CIPHER_find&type=code). What percentage of those projects depend on OSSL's signature vs. BSSL's signature? I don't know, but I would guess it would skew towards OSSL given its ubiquity. If we change the signature of `sk_SSL_CIPHER_find`, though, it would necessarily become a point of incompatibility with BSSL.